dev/benchmarks/macrobenchmarks/lib/src/web/bench_material_3_semantics.dart

```diff
@@ -0,0 +1,147 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter/scheduler.dart';
+import 'package:flutter/semantics.dart';
+
+import 'material3.dart';
+import 'recorder.dart';
+
+/// Measures the cost of semantics when constructing screens containing
+/// Material 3 widgets.
+class BenchMaterial3Semantics extends WidgetBuildRecorder {
+  BenchMaterial3Semantics() : super(name: benchmarkName);
+
+  static const String benchmarkName = 'bench_material3_semantics';
+
+  @override
+  Future<void> setUpAll() async {
+    FlutterTimeline.collectionEnabled = true;
+    super.setUpAll();
+    SemanticsBinding.instance.ensureSemantics();
+  }
+
+  @override
+  Future<void> tearDownAll() async {
+    FlutterTimeline.reset();
+  }
+
+  @override
+  void frameDidDraw() {
+    // Only record frames that show the widget. Frames that remove the widget
+    // are not interesting.
+    if (showWidget) {
+      final AggregatedTimings timings = FlutterTimeline.collect();
+      final AggregatedTimedBlock semanticsBlock = timings.getAggregated('SEMANTICS');
+      final AggregatedTimedBlock getFragmentBlock = timings.getAggregated('Semantics.GetFragment');
+      final AggregatedTimedBlock compileChildrenBlock = timings.getAggregated('Semantics.compileChildren');
+      profile!.addTimedBlock(semanticsBlock, reported: true);
+      profile!.addTimedBlock(getFragmentBlock, reported: true);
+      profile!.addTimedBlock(compileChildrenBlock, reported: true);
+    }
+
+    super.frameDidDraw();
+    FlutterTimeline.reset();
+  }
+
+  @override
+  Widget createWidget() {
+    return const SingleColumnMaterial3Components();
+  }
+}
+
+/// Measures the cost of semantics when scrolling screens containing Material 3
+/// widgets.
+///
+/// The implementation uses a ListView that jumps the scroll position between
+/// 0 and 1 every frame. Such a small delta is not enough for lazy rendering to
+/// add/remove widgets, but its enough to trigger the framework to recompute
+/// some of the semantics.
+///
+/// The expect output numbers of this benchmarks should be very small as
+/// scrolling a list view should be a matter of shifting some widgets and
+/// updating the projected clip imposed by the viewport. As of June 2023, the
+/// numbers are not great. Semantics consumes >50% of frame time.
```

Comment:
It seems like the 50% ratio is more important than the absolute numbers. Should we report the ratio instead? i.e. run the same widget with semantics and without semantics, and report the ratio between the two.

Reply:
I actually attempted it just to rediscover that our benchmark system is designed to only accept timeline data in microseconds. Maybe we should generalize it to accept other units.